scarb/src/ops/package.rs

```diff
@@ -0,0 +1,194 @@
+use std::collections::BTreeMap;
+
+use anyhow::{ensure, Result};
+use camino::Utf8PathBuf;
+
+use scarb_ui::components::Status;
+
+use crate::core::{Package, PackageId, PackageName, Workspace};
+use crate::flock::FileLockGuard;
+use crate::{ops, DEFAULT_SOURCE_PATH, MANIFEST_FILE_NAME};
+
+const VERSION: u8 = 1;
+const VERSION_FILE_NAME: &str = "VERSION";
+const ORIGINAL_MANIFEST_FILE_NAME: &str = "Scarb.orig.toml";
+
+const RESERVED_FILES: &[&str] = &[VERSION_FILE_NAME, ORIGINAL_MANIFEST_FILE_NAME];
+
+pub struct PackageOpts;
+
+/// A listing of files to include in the archive, without actually building it yet.
+///
+/// This struct is used to facilitate both building the package, and listing its contents without
+/// actually making it.
+type ArchiveRecipe = Vec<ArchiveFile>;
+
+struct ArchiveFile {
+    /// The relative path in the archive (not including top-level package name directory).
+    path: Utf8PathBuf,
+    #[allow(dead_code)]
+    /// The contents of the file.
+    contents: ArchiveFileContents,
+}
+
+enum ArchiveFileContents {
+    /// Absolute path to the file on disk to add to the archive.
+    OnDisk(Utf8PathBuf),
+
+    /// Generate file contents automatically.
+    ///
+    /// This variant stores a closure, so that file generation can be deferred to the very moment
+    /// it is needed.
+    /// For example, when listing package contents, we do not have files contents.
+    Generated(Box<dyn FnOnce() -> Result<Vec<u8>>>),
+}
+
+pub fn package(
+    packages: &[PackageId],
+    opts: &PackageOpts,
+    ws: &Workspace<'_>,
+) -> Result<Vec<FileLockGuard>> {
+    before_package(ws)?;
+
+    packages
+        .iter()
+        .map(|pkg| {
+            let pkg_name = pkg.to_string();
+            let message = Status::new("Packaging", &pkg_name);
+            if packages.len() <= 1 {
+                ws.config().ui().verbose(message);
+            } else {
+                ws.config().ui().print(message);
+            }
+
+            package_one_impl(*pkg, opts, ws)
+        })
+        .collect()
+}
+
+pub fn package_one(
```

Comment:
Do we need this? I do not think this is used.

Reply:
Right, I was too much forward-thinking (this may be useful for `scarb publish` 😅). I'll remove it.